[Hao Sitong] iP

o-ohst opened the pull request:
# Karen Bot

**User review:**
> ⭐️⭐️⭐️⭐️⭐️ "This bot is amazing! Since using Karen Bot, I no longer have any todos or deadlines, cuz I no longer have a job!" -Damith

**Features:**
- simple
- lightweight
- unnecessary

How to use:

1. Go to [this link](https://www.youtube.com/watch?v=dQw4w9WgXcQ)

Instructions to get started:

- [ ] Delete this app
- [ ] Don't come back again

Run this `main` method to get started:
```java
public static void main (String[] args) {
    System.out.println("Hello world");
}
```

Commit:
Add Gradle support

Commit:
greet, echo, exit

Commit:
add, list

Commit:
mark as done

Commit:
todo, event, deadline

Commit:
automated text ui testing

Commit:
handle errors

Commit:
delete

Commit:
change runtest.sh

Commit:
save

Commit:
dates and times

Commit:
merge

Commit:
date and time

Commit:
more OOP

Commit:
packages

NicholasTYD commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,30 @@
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task{
+    protected LocalDateTime by;
```
```suggestion
    protected LocalDateTime dueDate;
```
Perhaps it would be better if you use dueDate instead of by? That could make field sound more intuitive especially when its referenced in other parts of the code.

NicholasTYD commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,42 @@
+        return this.taskList;
+    }
+    
+    public int size() {
```
```suggestion
    public int getTaskCount() {
```
Would it be better to rename this method to sound more like a verb? It may be more easily understood when calling this method from another class.

NicholasTYD commented on src/main/java/duke/Todo.java:
```diff
@@ -0,0 +1,24 @@
+package duke;
+
+public class Todo extends Task{
+    protected String by;
```
Is this field used? It might be better to remove it if its not needed to reduce unnecessary code.

NicholasTYD commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,66 @@
+    }
+    
+    public void sendEmptyTaskMessage() {
+        System.out.println(divider);
+        System.out.println("Empty task? Are you kidding me??");
+        System.out.println(divider);
```
```suggestion
        String s = divider + "\n" 
                + "Empty task? Are you kidding me??" + "\n" 
                + divider;
        System.out.println(s);
```
Would it be better to combine all the messages into one string before printing it out? It may reduce the need to call System.out.println() many times.

NicholasTYD reviewed (commented):
Overall, I found your code easy to read for the most part except a few places
where you could use more meaningful names for your fields. I noted a few areas where you could write more concise code as well. Otherwise, your iP looks pretty great so far!

Commit:
Merge remote-tracking branch 'upstream/add-gradle-support'

Commit:
tests

Commit:
A-JUnit

Commit:
add javadoc

Commit:
add javadoc

Commit:
add coding standard

Commit:
add find

Commit:
add find

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-Level-9'

xiaobill8 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,157 @@
+            }
+            if (text.startsWith("mark ")) {
+                int index = Integer.parseInt(text.split(" ")[1]) - 1;
+                if (index < 0 || index >= tasks.size()) {
```
Hi Sitong, you could do this in a try catch block instead since we are dealing with ArrayIndexOutOfBoundsException here

xiaobill8 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,49 @@
+        String[] dateArr = date.split("/");
+        String time = dtStrings[1];
+
+        LocalDateTime dt = LocalDateTime.of(
```
Perhaps you could use the DateTimeFormatterBuilder to create a custom Datetime format to accept in parsing, rather than constructing your own datetime

xiaobill8 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.time.LocalDateTime;
+
+public abstract class Task {
```
I like how you encapsulated task into an abstract class

xiaobill8 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,157 @@
+        while (true) {
+            String text = input.nextLine();
+            if (text.equals("bye")) {
+                ui.sendByeMessage();
```
Perhaps you could encapsulate each of these commands into a Command object so that less of the logic is present in the main Duke.java code

xiaobill8 reviewed (commented):
Very concise and well-written code overall!

xiaobill8 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,49 @@
+        String time = dtStrings[1];
+
+        LocalDateTime dt = LocalDateTime.of(
+                Integer.parseInt(dateArr[2]),
```
Basic indentation should be 4 spaces instead of 8 here

xiaobill8 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,79 @@
+
+    /**
+     * Load tasks from storage or create new file.
+     * @return ArrayList of tasks
```
There should be an empty line between description and parameter/return section

xiaobill8 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    /**
+     * Save all tasks to storage
```
There should be an empty line between description and parameter/return section

xiaobill8 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,49 @@
+public class Parser {
+
+    /**
+     * Decodes a task from the given encoded string
```
There should be an empty line between description and parameter/return section

xiaobill8 commented on src/test/java/duke/TodoTest.java:
```diff
@@ -0,0 +1,13 @@
+
+public class TodoTest {
+    @Test
+    public void newTodoCreated(){
```
Perhaps you could name this as newTodoCreationTest()

xiaobill8 reviewed (commented):
Sorry, figured that the review should be on coding standards. Generally well done! Just take note of indentation and empty line between JavaDoc description and param/return section.

Commit:
add GUI

Commit:
add GUI

Commit:
Use assertions to check for valid datetime inputs.

Currently, bad inputs are not checked and can go unnoticed.

Assertions can help to detect bad inputs properly and early.

Assertions are used temporarily for simplicity.

Commit:
Improve code quality by making file path a constant.

Currently, the file path is hard coded.

Making it static and final allows it to be a single, constant value, which is appropriate for a file path.

o-ohst committed:
Merge branch 'master' into branch-A-Assertions

o-ohst committed:
Merge pull request #1 from o-ohst/branch-A-Assertions

add assertions

o-ohst committed:
Merge pull request #2 from o-ohst/branch-A-CodeQuality

Branch a code quality

Commit:
snooze

Commit:
level 10

Commit:
user guide

Commit:
user guide
